### ISSUE TITLE
depends_on keyword not working as expected

### ISSUE THREAD

**Issue by psfinaki** [CompositionalIT/farmer#446]
Well basically [this](https://compositionalit.github.io/farmer/api-overview/dependencies/#setting-dependencies) code just doesn't compile. 

But the whole docs section is confusing. For example, it says 
> you will still **need** to set dependencies when you have a relationship between builders, such as setting the key of a storage account on a web app.` 

Which I guess is not correct actually, things seem to work fine without the keyword.

Looks like there are issues at least for a month, see this [blog post](https://www.svenmalvik.com/azure-farmer-parameters/).

**Comment by isaacabraham** [comment]
This is probably just a case of documentation being out of date. In the earlier versions of Farmer, you explicitly had to supply `depends_on` all the time. In more recent releases, wherever we can, we infer the owner of any expressions and automatically set that resource as a dependency.

I'll have a look through the blog post though - obviously we're not monitoring every blog out there so thank you for pointing it out to us.

It's also possibly that there is a breaking change in either removal of an overload for `depends_on`, or an issue in overload resolution across generic lists (which F# is quite strict on).

**Comment by psfinaki** [comment]
Awesome, yes, it's probably about documentation and just coming up with a case when the necessary dependency would not be inferred and one would actually have to specify it.

**Comment by isaacabraham** [comment]
@psfinaki I've updated the docs around dependencies and pushed (and I might have found a nice solution for the list situation as well).

The reason that this code was valid originally was that in the earlier incarnations of Farmer, we had explicit overloads for `depends_on` for *each builder*. Obviously as the list grew, this quickly became unmaintainable and we moved to a plugin model through the `IBuilder` interface. However, doing this removed the ability to add a list of arbitrary builders without upcasting to IBuilder. This is a restriction in F# (one that I wish would be removed) but I'm hopeful we can remove it in future.
